test/algorithms/classifiers/test_vqc.py

```diff
@@ -51,351 +67,222 @@ def setUp(self):
-
-        num_qubits, feature_map, ansatz = None, None, None
-
-        if provide_num_qubits:
-            num_qubits = num_inputs
-        if provide_feature_map:
-            feature_map = ZZFeatureMap(num_inputs)
-        if provide_ansatz:
-            ansatz = RealAmplitudes(num_inputs, reps=1)
-
-        classifier = VQC(
-            num_qubits=num_qubits,
-            feature_map=feature_map,
-            ansatz=ansatz,
-            quantum_instance=self.qasm_quantum_instance,
+        multiclass_x, multiclass_y = make_classification(
+            n_samples=9,
+            n_features=self.num_features,
+            n_classes=3,
+            n_redundant=0,
+            n_clusters_per_class=1,
+            class_sep=5.0,
+            random_state=algorithm_globals.random_seed,
+        )
+        self.multiclass_dataset = _Dataset(
+            "multiclass", MinMaxScaler().fit_transform(multiclass_x), _one_hot_encode(multiclass_y)
         )
 
-        # construct data
-        num_samples = 5
-        # pylint: disable=invalid-name
-        X = algorithm_globals.random.random((num_samples, num_inputs))
-        y = 1.0 * (np.sum(X, axis=1) <= 1)
-        while len(np.unique(y)) == 1:
-            X = algorithm_globals.random.random((num_samples, num_inputs))
-            y = 1.0 * (np.sum(X, axis=1) <= 1)
-        y = np.array([y, 1 - y]).transpose()  # VQC requires one-hot encoded input
-
-        # fit to data
-        classifier.fit(X, y)
-
-        # score
-        score = classifier.score(X, y)
-        self.assertGreater(score, 0.5)
+        self.datasets = [self.binary_dataset, self.multiclass_dataset]
+        self.num_classes_by_batch = []
 
-    @data(
-        # optimizer, quantum instance
-        ("cobyla", "statevector"),
-        ("cobyla", "qasm"),
-        ("bfgs", "statevector"),
-        ("bfgs", "qasm"),
-        (None, "statevector"),
-        (None, "qasm"),
-    )
-    def test_multiclass(self, config):
-        """Test multiclass VQC."""
-        opt, q_i = config
-
-        if q_i == "statevector":
-            quantum_instance = self.sv_quantum_instance
-        elif q_i == "qasm":
-            quantum_instance = self.qasm_quantum_instance
-        else:
-            quantum_instance = None
-
-        if opt == "bfgs":
-            optimizer = L_BFGS_B(maxiter=5)
-        elif opt == "cobyla":
-            optimizer = COBYLA(maxiter=25)
-        else:
-            optimizer = None
-
-        num_inputs = 2
-        feature_map = ZZFeatureMap(num_inputs)
-        ansatz = RealAmplitudes(num_inputs, reps=1)
-        # fix the initial point
-        initial_point = np.array([0.5] * ansatz.num_parameters)
-
-        # construct classifier - note: CrossEntropy requires eval_probabilities=True!
+    def test_VQC(self):
+        """
+        Test VQC with binary and multiclass data using a range of quantum
+        instances, numbers of qubits, feature maps, and optimizers.
+        """
+
+        for (
```

Comment:
Since setup is called ahead of the test, things like the quantum_instances etc will be re-used in this for loop. Sure the whole test, that has the for loop, will be reproducible, but subsequent loops in the for will be re-using any objects. E.g if you had the same parameterization in two different tests you might not get the exact same result because of the random sampling in the qasm simulator would not have been reseeded. Maybe overall here that's not too much of a concern here. With the ddt decorator the test will have setup run for each parameterization; i.e. for each loop of this test so they would be independent of one another.

Reply:
Ah, right. I guess I could call setUp() and tearDown() manually at the start and end of the subTest, respectively. Though that doesn't seem all that elegant either. Perhaps it would be best to go back to DDT then...

Reply:
As it was a quick fix, I added the setUp and tearDown calls where they're required to ensure consistency. @adekusar-drl, do you have a preference for using DDT or subTests as I have?

Reply:
Personally, I don't mind having both ddt and subTests, but considering that we have a lot of ddt tests, I think, it is better to have ddt for consistency.

Reply:
Sure, no problem. There are some advantages to using DDT. For example, one can run a specific data-generated test individually without having to loop over all of them, which is nice. For some reason, I didn't like calling a function with String arguments to instantiate the immutable classes, but it seems this is the best way to do things. I'll refactor it again ASAP.

Reply:
Ok, I've refactored it in the first way that you suggested above, @woodsp-ibm.

> You could define functions, say on the test class, and have these be called by ddt to create the instances on demand. You could also do it in setup like before but say place the instances into a dictionary and put the keys in ddt to lookup the instance you want in the test - these could be string keys. This would then be a simple key lookup rather than the if tests, based on the strings, that were there before.

Hopefully, I implemented it correctly this time :sweat_smile:.

Reply:
Great, thanks for hanging in there with this.